### PR TITLE
Fix parser/parser state issues with removing/renaming components

### PR DESCRIPTION
**Pull request by Hosch250**

### CODE REVIEW

File: Rubberduck.Parsing/VBA/RubberduckParserState.cs [hunk 1]
```diff
@@ -430,7 +430,7 @@ public void AddDeclaration(Declaration declaration)
             }
         }
 
-        public void ClearStateCache(VBProject project)
+        public void ClearStateCache(VBProject project, bool setStateChanged = false)
```
Comment (retailcoder) [inline Rubberduck.Parsing/VBA/RubberduckParserState.cs]: `setStateChanged` isn't an ideal wording; we're not "setting" anything here, rather *notifying* whoever is listening for parser state events, that the state has changed.

I would have gone with `notifyStateChanged` I think.

File: Rubberduck.Parsing/VBA/RubberduckParserState.cs [hunk 2]
```diff
@@ -458,7 +461,7 @@ public void ClearBuiltInReferences()
             }
         }
 
-        public bool ClearStateCache(VBComponent component)
+        public bool ClearStateCache(VBComponent component, bool setStateChanged = false)
```
Comment (retailcoder) [inline Rubberduck.Parsing/VBA/RubberduckParserState.cs]: Ditto here

File: Rubberduck.Parsing/VBA/RubberduckParserState.cs [hunk 3]
```diff
@@ -507,10 +508,56 @@ public bool ClearStateCache(VBComponent component)
...
+            if (setStateChanged)
+            {
+                OnStateChanged();
+            }
+
             Debug.WriteLine("ClearDeclarations({0}): {1} - {2} declarations removed", component.Name, success ? "succeeded" : "failed", declarationsRemoved);
             return success;
         }
 
+        public bool RemoveRenamedComponent(VBComponent component, string oldComponentName)
```
Comment (retailcoder) [inline Rubberduck.Parsing/VBA/RubberduckParserState.cs]: I don't like the copy-paste going on here, this isn't trivial logic, and it's too easy to modify one place and miss the other.

File: Rubberduck.VBEEditor/QualifiedModuleName.cs [hunk 4]
```diff
@@ -60,6 +60,34 @@ public QualifiedModuleName(VBComponent component)
         }
 
         /// <summary>
+        /// Creates a QualifiedModuleName for removing renamed declarations.
+        /// Do not use this overload.
+        /// </summary>
+        public QualifiedModuleName(VBComponent component, string oldComponentName)
```
Comment (retailcoder) [inline Rubberduck.VBEEditor/QualifiedModuleName.cs]: There was no way of chaining the constructors?

File: Rubberduck.Parsing/VBA/RubberduckParserState.cs (hunk 1 above)
Comment (Hosch250) [inline Rubberduck.Parsing/VBA/RubberduckParserState.cs]: Sounds like a good name.

File: Rubberduck.Parsing/VBA/RubberduckParserState.cs (hunk 3 above)
Comment (Hosch250) [inline Rubberduck.Parsing/VBA/RubberduckParserState.cs]: I'll extract it to another method.

File: Rubberduck.VBEEditor/QualifiedModuleName.cs (hunk 4 above)
Comment (Hosch250) [inline Rubberduck.VBEEditor/QualifiedModuleName.cs]: I'll check, but I don't think so.

Comment (Hosch250) [inline Rubberduck.VBEEditor/QualifiedModuleName.cs]: No.  `QualifiedModuleName(VBProject project)` leaves the component name null.  `public QualifiedModuleName(VBComponent component)` gets the name from the component, which would be the new name by this point.  `QualifiedModuleName(string projectName, string projectPath, string componentName)` does not calculate the correct project id.

**Comment by Hosch250** [comment]
Thanks.  I'll take a look.

**Pull request merged**